string/Cursor.ts

```diff
@@ -0,0 +1,61 @@
+/** Describes a string and an index within it. */
+export class Cursor {
+  #input: { value: string };
+
+  constructor(
+    input: string,
+    public index: number = 0,
+  ) {
+    this.#input = { value: input };
+  }
+
+  /** Create a new Cursor with the same input and index. */
+  static copy(input: Cursor): Cursor {
+    const cursor = new Cursor("", input.index);
+    cursor.#input = input.#input;
+
+    return cursor;
+  }
+
+  /** An immutable reference to the input string. */
+  get input(): string {
+    return this.#input.value;
+  }
+
+  /** The character at the current index, if extant. */
+  get char(): string | undefined {
+    return this.#input.value.at(this.index);
+  }
+
+  /** The substring from the start of the string to the current index. */
+  get antecedent(): string {
```

Comment:
I looked all over for a better word than this - Copilot, GPT4, thesauruses... no dice. If you can think of something better I'm definitely interested. The issues with this one are that it's hard to spell correctly and it's an uncommon word.